codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/QueryShapeSerVisitor.java

```diff
@@ -0,0 +1,229 @@
+        return Collections.singletonMap("value", "query.Value");
+    }
+
+    @Override
+    protected void serializeCollection(GenerationContext context, CollectionShape shape) {
+        GoWriter writer = context.getWriter();
+        MemberShape member = shape.getMember();
+        Shape target = context.getModel().expectShape(member.getTarget());
+
+        writer.write("array := value.Array($S)", getSerializedLocationName(member, "member"));
+        writer.write("");
+
+        writer.openBlock("for i := range v {", "}", () -> {
+            writer.write("av := array.Value()");
+
+            // Null values should be omitted for query.
+            if (!target.hasTrait(EnumTrait.class)) {
+                writer.openBlock("if vv := v[i]; vv == nil {", "}", () -> {
+                    writer.write("continue");
+                });
+            }
+
+            target.accept(getMemberSerVisitor(shape.getMember(), "v[i]", "av"));
+        });
+        writer.write("return nil");
+    }
+
+    @Override
+    protected void serializeDocument(GenerationContext context, DocumentShape shape) {
+        LOGGER.warning("Document type is unsupported for Query serialization.");
+        context.getWriter().write("return &smithy.SerializationError{Err: fmt.Errorf("
+                + "\"Document type is unsupported for the query protocol.\")}");
+    }
+
+    @Override
+    protected void serializeMap(GenerationContext context, MapShape shape) {
+        GoWriter writer = context.getWriter();
+        Shape target = context.getModel().expectShape(shape.getValue().getTarget());
+
+        String keyLocationName = getSerializedLocationName(shape.getKey(), "key");
+        String valueLocationName = getSerializedLocationName(shape.getValue(), "value");
+        writer.write("object := value.Map($S, $S)", keyLocationName, valueLocationName);
+        writer.write("");
+
+        // Create a sorted list of the map's keys so we can have a stable body.
+        // Ideally this would be a function we dispatch to, but the lack of generics make
+        // that impractical since you can't make a function for a map[string]any
+        writer.write("keys := make([]string, 0, len(v))");
+        writer.write("for key := range v { keys = append(keys, key) }");
+        writer.addUseImports(GoDependency.standardLibraryDependency("sort", "1.14"));
+        writer.write("sort.Strings(keys)");
+        writer.write("");
+
+        writer.addUseImports(SmithyGoDependency.FMT);
+        writer.openBlock("for _, key := range keys {", "}", () -> {
+            writer.write("om := object.Key(key)");
+
+            // Null values should be omitted for query.
+            if (!target.hasTrait(EnumTrait.class)) {
+                writer.openBlock("if vv := v[key]; vv == nil {", "}", () -> {
+                    writer.write("continue");
+                });
+            }
+
+            target.accept(getMemberSerVisitor(shape.getValue(), "v[key]", "om"));
```

Comment:
That's a good question, I'll have to check.

Reply:
It looks like there isn't a representation for empty lists/maps. Wouldn't the for loop handle ignoring empty lists though?

Reply:
It should, I think my concern was mostly in the context if a prefix key was defined would an entry in the `url.Values` be created? If no, then probably don't need to add any additional checks.

Reply:
I don't *think* that'll be an issue since the collection types don't put anything into the list until you start adding entries. I'll double check.

Reply:
I updated it to put the entry object creation after the null check, which does actually fix this issue for maps. For lists it's a readability improvement. I also added a short-circuit check for maps to avoid the extra work/allocations done to sort them.

Reply:
I also added one for lists just to avoid creating the extra allocation